### ISSUE TITLE
RVect4 component w

### ISSUE THREAD

**Issue by TBog** [Rajawali/Rajawali#1858]
#### Rajawali Version or Branch
master

#### Device and Android Version
device SM-T705 android 5.0.2

#### Summary
RVect4 **w** component should return the same type of ShaderVar as **a** or **z** component

#### Steps to Reproduce
GL_POSITION.w().assign( GL_POSITION.z() );

#### Trace or Log Output
generated shader code for the above will look like this:
vec4 gl_Position.w = gl_Position.z;

and produce error:
[org.rajawali3d.materials.Material] Could not compile vertex shader:
Shader log: 0:29: L0006: Illegal identifier name 'gl_Position'
0:29: L0001: Expected token ';', found '.'

**Comment by jwoolston** [comment]
The shaderlog is telling you you cant define a new variable with the name gl_Position. This is because it is a reserved name in GLSL.

**Comment by TBog** [comment]
I am not defining, I am setting a value.
If I write `GL_POSITION.a().assign( GL_POSITION.z() );` it works as expected!

**Comment by jwoolston** [comment]
`vec4 gl_Position.w = gl_Position.z;` The fact that you have the data type at the start of the line shows you are defining.

**Comment by jwoolston** [comment]
Sorry, was early. The shader builder is generating the wrong GLSL and trying to declare `gl_Position` in that case rather than simple assignment (why the `vec4` shows up).

**Comment by contriteobserver** [comment]
using this test case to make the issue manifest
```
    private class ConstructedShaderRenderer extends Renderer {
        Object3D obj;

        public ConstructedShaderRenderer(Context context) {
            super(context);
            RajLog.setDebugEnabled(true);
        }

        @Override
        protected void initScene() {
            try {
                getCurrentScene().setBackgroundColor(Color.BLUE);

                VertexShader vertexShader = new VertexShader() {
                    private RMat4 uMVPMatrix;
                    private RVec4 aPosition;

                    @Override
                    public void initialize() {
                        super.initialize();

                        addPrecisionQualifier(DataType.FLOAT, Precision.MEDIUMP);
                        // -- uniforms
                        uMVPMatrix = (RMat4) addUniform(DefaultShaderVar.U_MVP_MATRIX);
                        // -- attributes
                        aPosition = (RVec4) addAttribute(DefaultShaderVar.A_POSITION);
                    }

                    @Override
                    public void main() {
                        RVec4 location = new RVec4("location");
                        location.assign(uMVPMatrix.multiply(aPosition));
                        //GL_POSITION.wxyz().assign(location.wxyz());
                        GL_POSITION.xyz().assign(location.xyz());
                        GL_POSITION.w().assign(location.w());
                    }
                };

                FragmentShader fragmentShader = new FragmentShader() {
                    @Override
                    public void initialize() {
                        super.initialize();

                        addPrecisionQualifier(DataType.FLOAT, Precision.MEDIUMP);
                    }

                    @Override
                    public void main() {
                        GL_FRAG_COLOR.assign("vec4(0.0, 1.0, 0.0, 1.0)");
                    }
                };

                Material customMat = new Material(vertexShader, fragmentShader);
                customMat.setColorInfluence(0);
                obj = new Cube(1);
                obj.setMaterial(customMat);
                getCurrentScene().addChild(obj);

                getCurrentCamera().setPosition(3, 3, 3);
                getCurrentCamera().setLookAt(obj.getPosition());
            } catch (Exception e) {
                Log.d(getLocalClassName(), e.toString());
            }

        }

        @Override
        public void onOffsetsChanged(float xOffset, float yOffset, float xOffsetStep, float yOffsetStep, int xPixelOffset, int yPixelOffset) {
        }

        @Override
        public void onTouchEvent(MotionEvent event) {
        }
    }
```

**Comment by jwoolston** [comment]
Resolved via #1888
